mackerel-plugin-solr/solr.go

```diff
@@ -0,0 +1,243 @@
+					for _, allowKey := range allowKeys {
+						s.Stats[core][allowKey+"_"+escapeSlash(key)] = v2[allowKey].(float64)
+					}
+				}
+			}
+		}
+	}
+}
+
+func (s *SolrPlugin) loadStatsMbeanQueryHandler(core string) error {
+	uri := s.BaseURL + "/" + core + "/admin/mbeans?stats=true&wt=json&cat=QUERYHANDLER"
+	for _, path := range queryHandlerPaths {
+		uri += fmt.Sprintf("&key=%s", url.QueryEscape(path))
+	}
+	stats, err := s.getStats(uri)
+	if err != nil {
+		return err
+	}
+	s.setStatsMbean(core, stats, queryHandlerStatKeys)
+	return nil
+}
+
+func (s *SolrPlugin) loadStatsMbeanCache(core string) error {
+	stats, err := s.getStats(s.BaseURL + "/" + core + "/admin/mbeans?stats=true&wt=json&cat=CACHE&key=filterCache&key=perSegFilter&key=queryResultCache&key=documentCache&key=fieldValueCache")
+	if err != nil {
+		return err
+	}
+	s.setStatsMbean(core, stats, cacheStatKeys)
+	return nil
+}
+
+func (s *SolrPlugin) loadStats() error {
+	s.Stats = map[string](map[string]float64){}
+
+	stats, err := s.getStats(s.BaseURL + "/admin/cores?wt=json")
+	if err != nil {
+		return err
+	}
+	s.Cores = []string{}
+	for core, values := range stats["status"].(map[string]interface{}) {
+		s.Cores = append(s.Cores, core)
+		s.Stats[core] = map[string]float64{}
+		err := s.loadStatsCore(core, values)
+		if err != nil {
+			return err
+		}
+		err = s.loadStatsMbeanQueryHandler(core)
+		if err != nil {
+			return err
+		}
+		err = s.loadStatsMbeanCache(core)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func escapeSlash(slashIncludedString string) (str string) {
+	str = strings.Replace(slashIncludedString, "/", "", -1)
+	return
+}
+
+// FetchMetrics interface for mackerelplugin
+func (s SolrPlugin) FetchMetrics() (map[string]float64, error) {
+	stat := make(map[string]float64)
+	for core, stats := range s.Stats {
+		for k, v := range stats {
+			stat[core+"_"+k] = v
+		}
+	}
+	return stat, nil
+}
+
+// GraphDefinition interface for mackerelplugin
+func (s SolrPlugin) GraphDefinition() map[string](mp.Graphs) {
+	graphdef := make(map[string](mp.Graphs))
+	var labelCore string
+
+	for _, core := range s.Cores {
+		labelCore = strings.Title(core)
+
+		graphdef[fmt.Sprintf("%s.%s.docsCount", s.Prefix, core)] = mp.Graphs{
+			Label: fmt.Sprintf("%s DocsCount", labelCore),
+			Unit:  "integer",
+			Metrics: [](mp.Metrics){
+				mp.Metrics{Name: core + "_numDocs", Label: "NumDocs", Diff: false},
+				mp.Metrics{Name: core + "_deletedDocs", Label: "DeletedDocs", Diff: false},
+			},
+		}
+
+		for _, key := range []string{"indexHeapUsageBytes", "segmentCount", "sizeInBytes"} {
+			metricLabel := strings.Title(key)
+			graphdef[fmt.Sprintf("%s.%s.%s", s.Prefix, core, key)] = mp.Graphs{
+				Label: fmt.Sprintf("%s %s", labelCore, metricLabel),
+				Unit:  "integer",
+				Metrics: [](mp.Metrics){
+					mp.Metrics{Name: core + "_" + key, Label: metricLabel, Diff: false},
+				},
+			}
+		}
+
+		for _, key := range queryHandlerStatKeys {
```

Comment:
This graph definition have not any "Label" parameters effect in roops. The cause remains unknown. Hmm...